File: artifacts/api-server/src/repositories/postgres/audit-repository.ts
import { db } from "@workspace/db";
import { auditLogsTable } from "@workspace/db";
import { eq, and, gte, lte, sql } from "drizzle-orm";
import { generateId } from "../../lib/id.js";
import type { IAuditRepository } from "../interfaces.js";

export class PostgresAuditRepository implements IAuditRepository {
  async findAll(params: { page?: number; pageSize?: number; userId?: string; action?: string; role?: string; dateFrom?: string; dateTo?: string }) {
    const page = params.page ?? 1;
    const pageSize = params.pageSize ?? 50;
    const offset = (page - 1) * pageSize;
    const conditions = [];
    if (params.userId) conditions.push(eq(auditLogsTable.userId, params.userId));
    if (params.action) conditions.push(eq(auditLogsTable.action, params.action));
    if (params.role) conditions.push(eq(auditLogsTable.role, params.role));
    if (params.dateFrom) conditions.push(gte(auditLogsTable.createdAt, new Date(params.dateFrom)));
    if (params.dateTo) conditions.push(lte(auditLogsTable.createdAt, new Date(params.dateTo)));
    const where = conditions.length > 0 ? and(...conditions) : undefined;
    const [rows, countResult] = await Promise.all([
      db.select().from(auditLogsTable).where(where).limit(pageSize).offset(offset).orderBy(sql`${auditLogsTable.createdAt} DESC`),
      db.select({ count: sql<number>`count(*)` }).from(auditLogsTable).where(where),
    ]);
    return { data: rows, total: Number(countResult[0].count), page, pageSize };
  }

  async create(data: any) {
    const id = generateId();
    const now = new Date();
    await db.insert(auditLogsTable).values({ ...data, id, createdAt: now });
    return id;
  }

  async findAll_export(params: { dateFrom?: string; dateTo?: string }) {
    const conditions = [];
    if (params.dateFrom) conditions.push(gte(auditLogsTable.createdAt, new Date(params.dateFrom)));
    if (params.dateTo) conditions.push(lte(auditLogsTable.createdAt, new Date(params.dateTo)));
    const where = conditions.length > 0 ? and(...conditions) : undefined;
    return db.select().from(auditLogsTable).where(where).orderBy(sql`${auditLogsTable.createdAt} DESC`);
  }
}

File: artifacts/api-server/Repositories/AuditRepository.cs
using Microsoft.EntityFrameworkCore;
using TimesheetApi.Data;
using TimesheetApi.Helpers;
using TimesheetApi.Models;

namespace TimesheetApi.Repositories;

public class AuditRepository : IAuditRepository
{
    private readonly AppDbContext _db;
    public AuditRepository(AppDbContext db) { _db = db; }

    public async Task<PaginatedResult<object>> FindAll(int page, int pageSize, string? userId,
        string? action, string? role, string? dateFrom, string? dateTo)
    {
        var offset = (page - 1) * pageSize;
        var query = _db.AuditLogs.AsQueryable();
        if (!string.IsNullOrEmpty(userId)) query = query.Where(a => a.UserId == userId);
        if (!string.IsNullOrEmpty(action)) query = query.Where(a => a.Action == action);
        if (!string.IsNullOrEmpty(role)) query = query.Where(a => a.Role == role);
        if (!string.IsNullOrEmpty(dateFrom) && DateTime.TryParse(dateFrom, out var df))
            query = query.Where(a => a.CreatedAt >= df);
        if (!string.IsNullOrEmpty(dateTo) && DateTime.TryParse(dateTo, out var dt))
            query = query.Where(a => a.CreatedAt <= dt);

        var total = await query.CountAsync();
        var rows = await query
            .OrderByDescending(a => a.CreatedAt)
            .Skip(offset).Take(pageSize)
            .ToListAsync();

        return new PaginatedResult<object>(rows.Select(ToObj).ToList(), total, page, pageSize);
    }

    public async Task<string> Create(Dictionary<string, object?> data)
    {
        var log = new AuditLog
        {
            Id = IdGenerator.NewId(),
            UserId = data.TryGetValue("userId", out var uid) ? uid?.ToString() ?? "" : "",
            UserName = data.TryGetValue("userName", out var un) ? un?.ToString() : null,
            Role = data.TryGetValue("role", out var r) ? r?.ToString() ?? "" : "",
            Action = data.TryGetValue("action", out var a) ? a?.ToString() ?? "" : "",
            EntityType = data.TryGetValue("entityType", out var et) ? et?.ToString() ?? "" : "",
            EntityId = data.TryGetValue("entityId", out var eid) ? eid?.ToString() : null,
            OldValue = data.TryGetValue("oldValue", out var ov) ? ov?.ToString() : null,
            NewValue = data.TryGetValue("newValue", out var nv) ? nv?.ToString() : null,
            IpAddress = data.TryGetValue("ipAddress", out var ip) ? ip?.ToString() : null,
            CreatedAt = DateTime.UtcNow,
        };
        _db.AuditLogs.Add(log);
        await _db.SaveChangesAsync();
        return log.Id;
    }

    public async Task<List<object>> FindAllExport(string? dateFrom, string? dateTo)
    {
        var query = _db.AuditLogs.AsQueryable();
        if (!string.IsNullOrEmpty(dateFrom) && DateTime.TryParse(dateFrom, out var df))
            query = query.Where(a => a.CreatedAt >= df);
        if (!string.IsNullOrEmpty(dateTo) && DateTime.TryParse(dateTo, out var dt))
            query = query.Where(a => a.CreatedAt <= dt);

        var rows = await query.OrderByDescending(a => a.CreatedAt).ToListAsync();
        return rows.Select(ToObj).ToList<object>();
    }

    private static object ToObj(AuditLog a) => new
    {
        id = a.Id,
        userId = a.UserId,
        userName = a.UserName,
        role = a.Role,
        action = a.Action,
        entityType = a.EntityType,
        entityId = a.EntityId,
        oldValue = a.OldValue,
        newValue = a.NewValue,
        ipAddress = a.IpAddress,
        createdAt = a.CreatedAt,
    };
}
